### PR TITLE
[RLlib] Remove `vtrace_drop_last_ts` option and add proper vf bootstrapping to IMPALA and APPO.

### PR DESCRIPTION
**Pull request by sven1977**
IMPALA and APPO both use the vtrace value bootstrapping method for better correcting for the slight off-policy'ness of their async samplers.

All of RLlib's implementations (tf/torch; Learner API-stack/old API stack) perform a "trick" to get around the need to compute an extra bootstrapped value at the very end of any sampled trajectory. PPO - on the other hand - performs this step correctly in its Policies' `postprocess_trajectory()` methods.
Instead, for APPO and IMPALA, RLlib cuts the last timesteps in each row of the BxT shaped train batch (on all data columns used in the loss, such as rewards, dones, etc..) and use that last timestep as the bootstrapped value. This might be problematic in cases, where an important reward is located exactly on the dropped timestep. Such a reward signal would simply be ignored by the algo.

An option to NOT cut this last timestep has recently been introduced (`config.vtrace_drop_last_ts=False`), however, the inaccuracy in not having a proper value-function computed V(t+1) at the end of a trajectory remains, b/c the core problem of this computation not happening in the `postprocess_trajectory` method has not been solved.

This PR therefore changes:
* The `vtrace_drop_last_ts` option is deprecated for all frameworks and API stacks.
* Proper value function bootstrapping is performed correctly at the end of each trajectory, using the next_obs as obs input, the last internal state output as internal state input (in case of RNNs), and storing the results in a new SampleBatch column, called VALUES_BOOTSTRAPPED. Hereby, the bootstrapped values are only stored on the last t of the batch (all other t's are set to 0.0), indicating that these values are to be used at t+1 (bootstrapping).
* In the APPO/IMPALA loss functions, we now properly utilize these bootstrapped values at the ends of the sampled (and possibly zero-padded) value tensors.

All the above steps fix the underlying mathematical inaccuracy and should yield better learning performance.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/evaluation/postprocessing.py [hunk 1]
```diff
@@ -179,10 +179,29 @@ def compute_gae_for_sample_batch(
...
+        use_gae=policy.config["use_gae"],
+        use_critic=policy.config.get("use_critic", True),
+    )
+
+    return batch
+
+
+@DeveloperAPI
+def compute_bootstrap_value(sample_batch, policy):
+    """TODO (sven): docstr"""
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/evaluation/postprocessing.py:200]: we should make it clear that we're modifying sample_batch in place.

File: rllib/evaluation/tests/test_trajectory_view_api.py [hunk 2]
```diff
@@ -129,7 +129,7 @@ def test_traj_view_lstm_prev_actions_and_rewards(self):
             view_req_policy = policy.view_requirements
             # 7=obs, prev-a + r, 2x state-in, 2x state-out.
             assert len(view_req_model) == 7, view_req_model
-            assert len(view_req_policy) == 22, (len(view_req_policy), view_req_policy)
+            assert len(view_req_policy) == 23, (len(view_req_policy), view_req_policy)
```
Comment (avnishn) [inline rllib/evaluation/tests/test_trajectory_view_api.py:132]: Did this change by 1 because we added a new sample_batch key?

**Review by avnishn — COMMENTED** [review]
I have questions about why we need to concatenate together values_time_major and bootstrap_values_time major, if we end up passing them separately to the vtrace function. 

Also I noticed that you changed APPO learner, but didn't make any modifications to the IMPALA learner. Was this intended? It seems like it may have not been since you also modified the impala policies.

File: rllib/evaluation/tests/test_trajectory_view_api.py (hunk 2 above)
Comment (sven1977) [reply rllib/evaluation/tests/test_trajectory_view_api.py:132]: Yup, correct.

File: rllib/evaluation/postprocessing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/evaluation/postprocessing.py:200]: sorry, forgot, will fix.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/evaluation/postprocessing.py:200]: done

File: rllib/algorithms/appo/appo_tf_policy.py [hunk 3]
```diff
@@ -159,12 +158,28 @@ def make_time_major(*args, **kw):
             prev_action_dist = dist_class(behaviour_logits, self.model)
             values = self.model.value_function()
             values_time_major = make_time_major(values)
+            bootstrap_values_time_major = make_time_major(
+                train_batch[SampleBatch.VALUES_BOOTSTRAPPED]
+            )
+            # Add values to bootstrap values to yield correct t=1 to T+1 trajectories,
+            # with T being the rollout length (max trajectory len).
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/appo/appo_tf_policy.py:166]: I expanded the docstring of `ray.rllib.evaluation.postprocessing.compute_bootstrap_value()` and added the computation example there (how to add the two columns together to get the final vtrace-usable value estimates).
Then I linked from each of the 4 loss functions (IMPALA+APPO vs torch+tf) to this new docstring.

File: rllib/algorithms/appo/tf/appo_tf_learner.py [hunk 4]
```diff
@@ -61,17 +61,39 @@ def compute_loss_for_module(
...
+            [tf.zeros((1, B)), bootstrap_values_time_major], axis=0
+        )
+        # Note that the `SampleBatch.VALUES_BOOTSTRAPPED` values are always recorded
+        # ONLY at the last ts of a trajectory (for the following timestep,
+        # which is one past(!) the last ts). All other values in that tensor are
+        # zero.
+        # Adding values and bootstrap_values yields the correct values+bootstrap
+        # configuration, from which we can then take t=-1 (last timestep) to get
+        # the bootstrap_value arg for the vtrace function below.
+        values_time_major += bootstrap_values_time_major
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/appo/tf/appo_tf_learner.py:96]: I do think we still need this. For the following reason:
In APPO and IMPALA, we build/preprocess the train batch along the rollout_fragment_len (or `max_seq_len` if LSTM) boundaries. This means that in case a rollout ends within an episode, this rollout's last trajectory will end up in the train batch with a zero-padded right side, thus, the bootstrapped value for this fragment is in the middle of the train batch, NOT at time-axis index -1!
So adding these two together here covers that particular case as well. It'll lead to most bootstrapped value to be at the end of the train batch rows, but some (in those cases where the rollout was done within an episode) will be located in the middle of the train batch's rows. Here: "row" means a trajectory (along the T-axis) within the (B, T, ...) train batch.

File: rllib/algorithms/impala/impala.py [hunk 5]
```diff
@@ -293,6 +283,9 @@ def training(
         Returns:
             This updated AlgorithmConfig object.
         """
+        if vtrace_drop_last_ts is not None:
+            deprecation_warning(old="vtrace_drop_last_ts", error=False)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/impala/impala.py:287]: Agreed, forced explicitness is better here!

File: rllib/algorithms/impala/impala_tf_policy.py [hunk 6]
```diff
@@ -363,30 +382,21 @@ def make_time_major(*args, **kw):
...
+                dones=make_time_major(dones),
+                behaviour_action_logp=make_time_major(behaviour_action_logp),
+                behaviour_logits=make_time_major(unpacked_behaviour_logits),
+                target_logits=make_time_major(unpacked_outputs),
                 discount=self.config["gamma"],
-                rewards=make_time_major(rewards, drop_last=drop_last),
-                values=make_time_major(values, drop_last=drop_last),
-                bootstrap_value=make_time_major(values)[-1],
+                rewards=make_time_major(rewards),
+                values=values_time_major[:-1],
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/impala/impala_tf_policy.py:395]: True, but see my explanation above for those cases where a rollout ends in the middle of an episode and the bootstrapped value will then "show up" in the middle of the train batch row, NOT at -1.

File: rllib/algorithms/appo/tf/appo_tf_learner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/appo/tf/appo_tf_learner.py:98]: I'm struggling to understand how `compute_bootstrap_value` handles this situation where the rollout ends within an episode.  I'm reading the code for it, and on the surface I don't see anything that searches in the middle of an episode for the reward at the terminated timestep. It looks like we're only checking the last timestep in the sample batch.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/appo/tf/appo_tf_learner.py]: Note that `compute_bootstrap_value` is only ever called by a Policy's postprocess_trajectory method, which - by design - is only called when a rollout has ended (either within or at the terminal of an episode).
If at a terminal: Assume the value to be 0.0 (no value computation necessary)
If NOT at a terminal: Use the Policy's vf to compute the value at the last timestep of the trajectory. This is the "bootstrap" value to be used in the losses (instead of dropping the last ts and using that ts as a "bootstrapped" value).

File: rllib/algorithms/appo/appo_tf_policy.py [hunk 7]
```diff
@@ -203,26 +201,20 @@ def reduce_mean_valid(t):
...
                         behaviour_policy_logits=make_time_major(
-                            unpacked_behaviour_logits, drop_last=drop_last
+                            unpacked_behaviour_logits
                         ),
                         target_policy_logits=make_time_major(
-                            unpacked_old_policy_behaviour_logits, drop_last=drop_last
-                        ),
-                        actions=tf.unstack(
-                            make_time_major(loss_actions, drop_last=drop_last), axis=2
+                            unpacked_old_policy_behaviour_logits
```
Comment (avnishn) [inline rllib/algorithms/appo/appo_tf_policy.py:207]: do you not need to keep the tf unstack logic? I guess if vtrace tests are passing, then no...

File: rllib/algorithms/impala/impala_torch_policy.py [hunk 8]
```diff
@@ -222,6 +222,8 @@ def __init__(self, observation_space, action_space, config):
             max_seq_len=config["model"]["max_seq_len"],
         )
 
+        ValueNetworkMixin.__init__(self, config)
```
Comment (avnishn) [inline rllib/algorithms/impala/impala_torch_policy.py:225]: why did we add the value network mixin here? what were we doing before?

File: rllib/evaluation/postprocessing.py [hunk 9]
```diff
@@ -217,18 +270,18 @@ def compute_gae_for_sample_batch(
...
-    # Adds the policy logits, VF preds, and advantages to the batch,
-    # using GAE ("generalized advantage estimation") or not.
-    batch = compute_advantages(
-        sample_batch,
-        last_r,
-        policy.config["gamma"],
-        policy.config["lambda"],
-        use_gae=policy.config["use_gae"],
-        use_critic=policy.config.get("use_critic", True),
+    # Set the SampleBatch.VALUES_BOOTSTRAPPED field to VF_PREDS[1:] + the
```
Comment (avnishn) [inline rllib/evaluation/postprocessing.py:273]: why is it not VF_PREDS[:-1]

**Review by avnishn — APPROVED** [review]
small questions, mainly for my understanding. otherwise lgtm

File: rllib/evaluation/postprocessing.py (hunk 9 above)
Comment (sven1977) [reply rllib/evaluation/postprocessing.py:273]: Then we would drop the last ts again.
You want to have in the "bootstrapped sequence" exactly the values of t+1 (for each t). Such that bootstrapped[-1] is then the actually bootstrapped value estimate.

File: rllib/algorithms/impala/impala_torch_policy.py (hunk 8 above)
Comment (sven1977) [reply rllib/algorithms/impala/impala_torch_policy.py:225]: For IMPALA, nothing. We did not need to compute any VF_PREDs, ever. Because we were doing this in the loss function, using a `model.value()` call.

File: rllib/evaluation/postprocessing.py (hunk 9 above)
Comment (avnishn) [reply rllib/evaluation/postprocessing.py:273]: oh wait this is a part of the impala math, my bad.
